src/osp/scientific/kepler.h

```diff
@@ -0,0 +1,161 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2022 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "osp/core/math_types.h"
+
+#include <optional>
+
+namespace osp
+{
+
+/**
+ * @brief Holds parameters to fully define a Keplerian (ideal 2-body) orbit
+ */
+struct KeplerOrbit
+{
+private:
+    static constexpr double KINDA_SMALL_NUMBER = 1.0E-8;
+    // These are private as we may want to change the internal representation in the future
+    double m_semiMajorAxis;
+    double m_eccentricity;
+    double m_inclination;
+    double m_argumentOfPeriapsis;
+    double m_longitudeOfAscendingNode;
+    double m_meanAnomalyAtEpoch;
+    double m_epoch;
+    double m_r0;
+    double m_v0;
+    double m_h;
+
+    union
+    {
+        double m_F0;
+        double m_E0;
+    };
+
+    // Gravitational parameter of the central body (G*M)
+    double m_gravitationalParameter;
+
+    KeplerOrbit(double semiMajorAxis,
+                double eccentricity,
+                double inclination,
+                double argumentOfPeriapsis,
+                double longitudeOfAscendingNode,
+                double meanAnomalyAtEpoch,
+                double epoch,
+                double gravitationalParameter,
+                double r0,
+                double v0,
+                double m_E0,
+                double h) : m_semiMajorAxis(semiMajorAxis),
```

Comment:
The cool kids way of doing this would be
```cpp
KeplerOrbit(...)
 : m_semiMajorAxis(semiMajorAxis)
 , m_eccentricity(eccentricity)
 , m_inclination(inclination)
 ...
```

An alternative is that these parameters can be moved into a separate plain-old-data struct and stored as a private member. In this case, you only need to pass one thing through the constructor. you can drop the `m_` prefix and also initialize it with designated initializers, like `KeplerOrbit orbit{{semiMajorAxis = a, .eccentricity = b, ...}}`

Reply:
Could you send me an example of that second approach? That's some C++ magic that I haven't used before

Reply:
```cpp
// store this as a private member
struct KeplerOrbitParams
{
    double semiMajorAxis;
    double eccentricity;
    double inclination;
    double argumentOfPeriapsis;
    // pretend the rest are here
};

// constructor would look something like
KeplerOrbit::KeplerOrbit(KeplerOrbitParams params) : m_params{params} { };


// in some function, maybe in from_initial_conditions(...)?
// using Designated Initializer

KeplerOrbitParams params{
    .semiMajorAxis = semiMajorAxis,
    .eccentricity = e,
    .inclination = inclination,
    .argumentOfPeriapsis = w,
};

return KeplerOrbit{params};
```